File: expenseTracker/app/models/expense.rb
class Expense < ActiveRecord::Base

  def self.import(file)
    CSV.foreach(file.path, headers: true) do |row|
      hash = row.to_hash

      ## Normalize Data column names and data
      newHash = Hash.new
      newHash["date"] = Date::strptime(hash['date'], "%m/%d/%Y")
      newHash["category"] = hash["category"]
      newHash["employee_name"] = hash["employee name"]
      newHash["employee_addr"] = hash["employee address"]
      newHash["description"] = hash["expense description"]
      newHash["amount"] = hash["pre-tax amount"].to_d
      newHash["tax_type"] = hash["tax name"]
      newHash["tax"] = hash["tax amount"].to_d
      newHash["total_amount"] = (hash["pre-tax amount"].to_d + hash["tax amount"].to_d)

      Expense.create(newHash)
    end
  end

  def self.monthly_report
    expenses= Expense.all
    monthly_expense = Hash.new
    expenses.each do |expense|
      if monthly_expense.has_key?(expense.date.year) then
        if monthly_expense[expense.date.year].has_key?(expense.date.month) then
          monthly_expense[expense.date.year][expense.date.month] += expense.total_amount
        else
          monthly_expense[expense.date.year][expense.date.month] = 0
          monthly_expense[expense.date.year][expense.date.month] += expense.total_amount
        end
      else
        monthly_expense[expense.date.year] = Hash.new
        if monthly_expense[expense.date.year].has_key?(expense.date.month) then
          monthly_expense[expense.date.year][expense.date.month] += expense.total_amount
        else
          monthly_expense[expense.date.year][expense.date.month] = 0
          monthly_expense[expense.date.year][expense.date.month] += expense.total_amount
        end
      end
    end
    return monthly_expense
  end

end

File: app/repositories/expense.py
# -*- encoding: utf-8 -*-
import csv
from datetime import datetime

from app import db
from app.models.category import Category
from app.models.employee import Employee
from app.models.expense import Expense
from app.models.tax import Tax


class ExpenseRepository():
    """Repository to create and retrieve Expenses"""
    def from_csv(self, csvfile):
        categories = {}
        employees = {}
        taxes = {}

        data = csv.DictReader(csvfile)

        for row in data:
            if not row['category'] in categories:
                new_category = Category(row['category'])
                categories[row['category']] = new_category
                db.session.add(new_category)

            if not row['employee name'] in employees:
                new_employee = Employee(row['employee name'],
                                        row['employee address'])
                employees[row['employee name']] = new_employee
                db.session.add(new_employee)

            pre_tax_amount = float(row['pre-tax amount'].replace(',', ''))
            tax_amount = float(row['tax amount'].replace(',', ''))
            tax_percentage = tax_amount / pre_tax_amount

            if not row['tax name'] in taxes:
                new_tax = Tax(row['tax name'], tax_percentage)
                taxes[row['tax name']] = new_tax
                db.session.add(new_tax)

            # date, category, employee, tax, description, pre_tax_amount
            new_expense = Expense(
                datetime.strptime(row['date'], "%m/%d/%Y").date(),
                categories[row['category']],
                employees[row['employee name']],
                taxes[row['tax name']],
                row['expense description'],
                pre_tax_amount
            )
            db.session.add(new_expense)

        db.session.commit()
